src/lib/krad/packet.c

```diff
@@ -170,34 +180,24 @@ auth_generate_random(krb5_context ctx, uchar *rauth)
 /* Generate a response authenticator field. */
 static krb5_error_code
 auth_generate_response(krb5_context ctx, const char *secret,
-                       const krad_packet *response, const uchar *auth,
-                       uchar *rauth)
+                       const krb5_data *respbuf, const uint8_t *auth,
+                       uint8_t *rauth)
 {
-    krb5_error_code retval;
-    krb5_checksum hash;
-    krb5_data data;
-
-    /* Allocate the temporary buffer. */
-    retval = alloc_data(&data, response->pkt.length + strlen(secret));
-    if (retval != 0)
-        return retval;
+    krb5_crypto_iov input[4];
 
     /* Encoded RADIUS packet with the request's
      * authenticator and the secret at the end. */
-    memcpy(data.data, response->pkt.data, response->pkt.length);
-    memcpy(data.data + OFFSET_AUTH, auth, AUTH_FIELD_SIZE);
-    memcpy(data.data + response->pkt.length, secret, strlen(secret));
+    input[0].flags = KRB5_CRYPTO_TYPE_DATA;
+    input[0].data = make_data(respbuf->data, respbuf->length);
+    input[1].flags = KRB5_CRYPTO_TYPE_DATA;
+    input[1].data = make_data((uint8_t *)auth, AUTH_FIELD_SIZE);
+    input[2].flags = KRB5_CRYPTO_TYPE_DATA;
+    input[2].data = make_data((char *)secret, strlen(secret));
+    input[3].flags = KRB5_CRYPTO_TYPE_CHECKSUM;
+    input[3].data = make_data(rauth, AUTH_FIELD_SIZE);
 
     /* Hash it. */
-    retval = krb5_c_make_checksum(ctx, CKSUMTYPE_RSA_MD5, NULL, 0, &data,
-                                  &hash);
-    free(data.data);
-    if (retval != 0)
-        return retval;
-
-    memcpy(rauth, hash.contents, AUTH_FIELD_SIZE);
-    krb5_free_checksum_contents(ctx, &hash);
-    return 0;
+    return krb5_k_make_checksum_iov(ctx, CKSUMTYPE_RSA_MD5, NULL, 0, input, 4);
```

Comment:
I replaced `krb5_c_make_checksum()` by an iov checksum function to avoid allocating memory. I used `krb5_k_make_checksum_iov()` because `krb5_c_make_checksum_iov()` seems to not accept a NULL keyblock (it causes a segfault).